applets/notification_area/na-grid.c

```diff
@@ -84,17 +98,157 @@ compare_items (gconstpointer a,
+    }
+
+  /* only update item position if has changed from current */
+  gtk_container_child_get (GTK_CONTAINER (data->na_grid),
+                           item,
+                           "left-attach", &left,
+                           "width", &right,
+                           "top-attach", &top,
+                           "height", &bottom,
+                           NULL);
+  if (left != col || right != 1 || top != row || bottom != 1)
+    {
+      gtk_container_child_set (GTK_CONTAINER (data->na_grid),
+                               item,
+                               "left-attach", col,
+                               "width", 1,
+                               "top-attach", row,
+                               "height", 1,
+                               NULL);
+    }
+  /* increment to index of next item */
+  data->index++;
+}
+
+static void
+determine_grid_properties (NaGrid *self)
+{
+  gint rows, cols, length;
+  GtkOrientation orientation;
+  GtkAllocation allocation;
+  
+  orientation = gtk_orientable_get_orientation (GTK_ORIENTABLE (self));
+  gtk_widget_get_allocation (GTK_WIDGET (self), &allocation);
+  length = g_slist_length (self->items);
+
+  if (orientation == GTK_ORIENTATION_HORIZONTAL)
+    {
+      rows = MAX (1, allocation.height / self->min_icon_size);
+      cols = MAX (1, length / rows);
+      if (length % rows)
+        cols++;
+    } else {
+      cols = MAX (1, allocation.width / self->min_icon_size);
+      rows = MAX (1, length / cols);
+      if (length % cols)
+        rows++;
+    }
+  
+  self->grid_props->cols = cols;
+  self->grid_props->rows = rows;
+  self->grid_props->length = length;
+}
+
 static void
-reorder_items (GtkWidget *widget,
-               gpointer   user_data)
+refresh_grid (NaGrid *self)
 {
-  NaGrid *nb;
-  gint position;
+  PackData pack_data;
+  GSList *item;
+  GtkOrientation orientation;
+  gint row = 0, col = 0;
+  
+  g_return_if_fail (NA_IS_GRID (self));
+  
+  orientation = gtk_orientable_get_orientation (GTK_ORIENTABLE (self));
+  determine_grid_properties (self);
+
+  /*
+  * There might be a better way to do this
+  * The refresh is caused by all items being removed from the container first,
+  * then re-added at the position they're supposed to be at,
+  * then sorting them again in reorder_items_with_data()
+  *
+  * I couldn't find how to force the update with another mechanism?
+  * Just sorting them doesn't work.
+  * Removing them/re-adding them alone doesn't cause an update either.
+  * All these steps have to happen for the grid to update when the panel loading occurs, when a new item is added or when the min icon size is changed in the preferences...
+  */
+  for (item = self->items; item; item = item->next)
+    {
+      if (GTK_IS_WIDGET (item))
+        gtk_container_remove (GTK_CONTAINER (self), GTK_WIDGET (item));
+    }
 
-  nb = NA_GRID (user_data);
+  for (item = self->items; item; item = item->next)
+    {  
+      if (GTK_IS_WIDGET (item))
```

Comment:
Same here...

Reply:
My guess is that when you remove widget from container it is destroyed. Widgets have floating reference - https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#GInitiallyUnowned.

In short you need to temporarily increase reference to make sure it is not destroyed. Once it is added back you remove your temporarily reference.

Reply:
I've been testing, actually it seems all that code becomes unnecessary. since I have the size-allocate set on the grid as it should, the reordering works almost perfectly simply with reorder_items_with_data(). I guess that having the size-allocate on the panel widget, plus the notify::size-changed on the grid, just triggered many resizes that conflicted with each other?
Btw. since I'm not expert in gtk, what does issue "notify::size-changed" signals? is it necessary to have both, size-allocate and notify::size-changed?

Reply:
What is `notify::size-changed`? Quickly looking I don't see such property on `GtkGrid`, `GtkContainer` and/or `GtkWidget`...

Reply:
Ok I don't know where I picked that up.. I've been testing even orientation notify is now useless with size_allocate directly on the GtkGrid. I just removed all that now unnecessary code.